v2: Add guest allocator

rvolosatovs opened the pull request:
Part of PR series for #17

Closes https://github.com/enarx/sallyport/issues/35

Blocked by ~https://github.com/enarx/sallyport/pull/28~ ~#27 (due to `item` crate changes)~ #36

rvolosatovs commented on src/v2/src/guest/syscall/result/linux/x86_64.rs:
```diff
@@ -0,0 +1,78 @@
+
+impl<T> Result<T> {
+    #[inline]
+    pub const unsafe fn errno_unchecked(errno: c_int) -> Self {
```
This is necessary, because const options are unstable

haraldh commented:
Miri is failing with:

```
❯ cargo +nightly miri test
    Finished test [unoptimized + debuginfo] target(s) in 0.00s
     Running unittests (target/miri/x86_64-unknown-linux-gnu/debug/deps/sallyport-bed894b0ae9968c3)
warning: unused import: `core::ptr::null_mut`
  --> src/guest/syscall/result/linux/x86_64.rs:63:9
   |
63 |     use core::ptr::null_mut;
   |         ^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default


running 5 tests
test guest::syscall::result::linux::x86_64::tests::result ... ok
test guest::syscall::tests::exit ... error: Undefined Behavior: no item granting read access to tag <untagged> at alloc69380+0x48 found in borrow stack.
   --> /home/harald/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:701:9
    |
701 |         copy_nonoverlapping(src, tmp.as_mut_ptr(), 1);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item granting read access to tag <untagged> at alloc69380+0x48 found in borrow stack.
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
            
    = note: inside `std::ptr::read::<[usize; 2]>` at /home/harald/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:701:9
    = note: inside `std::ptr::const_ptr::<impl *const [usize; 2]>::read` at /home/harald/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/const_ptr.rs:810:18
note: inside `guest::alloc::output::OutRef::<[usize; 2]>::read::<guest::alloc::phase_alloc::Alloc<guest::alloc::phase_alloc::phase::Collect>>` at src/guest/alloc/output.rs:73:18
   --> src/guest/alloc/output.rs:73:18
    |
73  |         unsafe { self.as_ptr().read() }
    |                  ^^^^^^^^^^^^^^^^^^^^
note: inside `<guest::alloc::syscall::CommittedSyscall<guest::syscall::exit::Exit> as guest::alloc::Collect>::collect::<guest::alloc::phase_alloc::Alloc<guest::alloc::phase_alloc::phase::Collect>>` at src/guest/alloc/syscall.rs:116:36
   --> src/guest/alloc/syscall.rs:116:36
    |
116 |         T::collect(self.committed, self.ret_ref.read(col).into(), col)
    |                                    ^^^^^^^^^^^^^^^^^^^^^^
note: inside `guest::syscall::tests::Call::<guest::syscall::exit::Exit, (), 11_usize>::assert` at src/guest/syscall/tests.rs:55:25
   --> src/guest/syscall/tests.rs:55:25
    |
55  |         let collected = committed.collect(&alloc);
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `guest::syscall::tests::exit` at src/guest/syscall/tests.rs:62:5
   --> src/guest/syscall/tests.rs:62:5
    |
62  | /     Call::new(
63  | |         Exit { status: 2 },
64  | |         [
65  | |             SYSCALL_USIZE_COUNT * size_of::<usize>(),
...   |
79  | |     )
80  | |     .assert()
    | |_____________^
note: inside closure at src/guest/syscall/tests.rs:61:1
   --> src/guest/syscall/tests.rs:61:1
    |
60  |   #[test]
    |   ------- in this procedural macro expansion
61  | / fn exit() {
62  | |     Call::new(
63  | |         Exit { status: 2 },
64  | |         [
...   |
80  | |     .assert()
81  | | }
    | |_^

```

haraldh commented:
Also for miri I had to patch:
```diff
diff --git a/src/v2/src/guest/syscall/result/linux/x86_64.rs b/src/v2/src/guest/syscall/result/linux/x86_64.rs
index 9e6ce8c..48005f8 100644
--- a/src/v2/src/guest/syscall/result/linux/x86_64.rs
+++ b/src/v2/src/guest/syscall/result/linux/x86_64.rs
@@ -66,9 +66,14 @@ mod tests {
     fn result() {
         const ERRNO: c_int = libc::EPERM;
 
+        let mut buf: [u8; 0] = [];
+
         let expected: Result<isize> = unsafe { super::Result::errno_unchecked(ERRNO) };
         assert_eq!(
-            Result::from([unsafe { libc::read(-1, null_mut(), 0) } as usize, 0]),
+            Result::from([
+                unsafe { libc::read(-1, buf.as_mut_ptr() as _, 0) } as usize,
+                0
+            ]),
             expected
         );
 

```

rvolosatovs commented:
I changed the approach used in the test to prevent miri error regarding the syscall in the test.
I added a test for the allocator to reproduce the allocator error in a more isolated environment.

```
test guest::alloc::tests::alloc ... error: Undefined Behavior: no item granting read access to tag <untagged> at alloc66202+0x4 found in borrow stack.
    --> /nix/store/np81x8dbdl14lvddz8b2q6djyv9njb9x-rust-default-1.59.0-nightly-2021-12-22/lib/rustlib/src/rust/library/core/src/intrinsics.rs:2082:14
     |
2082 |     unsafe { copy_nonoverlapping(src, dst, count) }
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item granting read access to tag <untagged> at alloc66202+0x4 found in borrow stack.
     |
     = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
     = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
             
     = note: inside `std::intrinsics::copy_nonoverlapping::<u16>` at /nix/store/np81x8dbdl14lvddz8b2q6djyv9njb9x-rust-default-1.59.0-nightly-2021-12-22/lib/rustlib/src/rust/library/core/src/intrinsics.rs:2082:14
     = note: inside `std::ptr::const_ptr::<impl *const u16>::copy_to_nonoverlapping` at /nix/store/np81x8dbdl14lvddz8b2q6djyv9njb9x-rust-default-1.59.0-nightly-2021-12-22/lib/rustlib/src/rust/library/core/src/ptr/const_ptr.rs:889:18
note: inside `guest::alloc::output::OutRef::<u16>::copy_to::<guest::alloc::phase_alloc::Alloc<guest::alloc::phase_alloc::phase::Collect>, &mut u16>` at src/guest/alloc/output.rs:73:18
    --> src/guest/alloc/output.rs:73:18
     |
73   |         unsafe { self.as_ptr().copy_to_nonoverlapping(dest.borrow_mut(), 1) }
     |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `<guest::alloc::output::Output<u16, u16> as guest::alloc::Collect>::collect::<guest::alloc::phase_alloc::Alloc<guest::alloc::phase_alloc::phase::Collect>>` at src/guest/alloc/output.rs:187:9
    --> src/guest/alloc/output.rs:187:9
     |
187  |         self.data_ref.copy_to(col, self.val.borrow_mut());
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `guest::alloc::tests::alloc` at src/guest/alloc/tests.rs:108:16
    --> src/guest/alloc/tests.rs:108:16
     |
108  |     assert_eq!(out_u16.collect(&alloc), 0xfeed);
     |                ^^^^^^^^^^^^^^^^^^^^^^^
note: inside closure at src/guest/alloc/tests.rs:9:1
    --> src/guest/alloc/tests.rs:9:1
     |
8    |   #[test]
     |   ------- in this procedural macro expansion
9    | / fn alloc() {
10   | |     const USIZE_COUNT: usize = 16;
11   | |
12   | |     let mut buf = [usize::MAX; USIZE_COUNT];
...    |
120  | |     );
121  | | }
```

From what I understand, the issue stems from the fact that we are writing/reading to/from arbitrary pointers of unknown size.
I refactored the `Alloc` to hold `NonNull<[u8]>` rather than `*mut u8` and `free: usize` and use slice indexes to produce pointers (https://github.com/enarx/sallyport/pull/29/commits/d13f43c79929b07b6854f13674099feea75c2c1d), but miri still complains.
I suppose the issue is that `*mut [u8]` is converted to `*mut u8` and then to `*mut T` and miri complains about use of the resulting `*mut T`.
I am not sure that this is something we should worry about. It does not seem to be possible to allocate arbitrary `alloc::Layout`s without doing this.
@npmccallum @haraldh what do you think?
